// @sapphire/time-utilities/dist/index.js
"use strict";
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  Cron: () => import_cron.Cron,
  DEFAULT_SEPARATORS: () => import_duration.DEFAULT_SEPARATORS,
  DEFAULT_UNITS: () => import_duration.DEFAULT_UNITS,
  Duration: () => import_duration.Duration,
  DurationFormatter: () => import_duration.DurationFormatter,
  Time: () => import_duration.Time,
  TimeTypes: () => import_duration.TimeTypes,
  TimerManager: () => import_timer_manager.TimerManager,
  Timestamp: () => import_timestamp.Timestamp,
  allowedNum: () => import_cron.allowedNum,
  cronTokens: () => import_cron.cronTokens,
  days: () => import_timestamp.days,
  months: () => import_timestamp.months,
  partRegex: () => import_cron.partRegex,
  predefined: () => import_cron.predefined,
  tokens: () => import_timestamp.tokens,
  tokensRegex: () => import_cron.tokensRegex,
  wildcardRegex: () => import_cron.wildcardRegex
});
module.exports = __toCommonJS(src_exports);
var import_cron = require("@sapphire/cron");
var import_duration = require("@sapphire/duration");
var import_timer_manager = require("@sapphire/timer-manager");
var import_timestamp = require("@sapphire/timestamp");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Cron,
  DEFAULT_SEPARATORS,
  DEFAULT_UNITS,
  Duration,
  DurationFormatter,
  Time,
  TimeTypes,
  TimerManager,
  Timestamp,
  allowedNum,
  cronTokens,
  days,
  months,
  partRegex,
  predefined,
  tokens,
  tokensRegex,
  wildcardRegex
});
